core/src/main/java/com/amplitude/core/IdentityStore.kt

```diff
@@ -0,0 +1,178 @@
+package com.amplitude.core
+
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+internal const val ID_OP_SET = "\$set"
+internal const val ID_OP_UNSET = "\$unset"
+internal const val ID_OP_SET_ONCE = "\$setOnce"
+internal const val ID_OP_ADD = "\$add"
+internal const val ID_OP_APPEND = "\$append"
+internal const val ID_OP_PREPEND = "\$prepend"
+internal const val ID_OP_CLEAR_ALL = "\$clearAll"
+
+typealias IdentityListener = (Identity) -> Unit
+
+data class Identity(
+    val userId: String? = null,
+    val deviceId: String? = null,
+    val userProperties: Map<String, Any?> = mapOf(),
+) {
+    fun isUnidentified(): Boolean {
+        return userId.isNullOrEmpty() && deviceId.isNullOrEmpty()
+    }
+}
+
+interface IdentityStore {
+
+    interface Editor {
+
+        fun setUserId(userId: String?): Editor
+        fun setDeviceId(deviceId: String?): Editor
+        fun setUserProperties(userProperties: Map<String, Any?>): Editor
+        fun updateUserProperties(actions: Map<String, Map<String, Any?>>): Editor
+        fun commit()
+    }
+
+    fun editIdentity(): Editor
+    fun setIdentity(identity: Identity)
+    fun getIdentity(): Identity
+    fun addIdentityListener(listener: IdentityListener)
+    fun removeIdentityListener(listener: IdentityListener)
+}
+
+internal class IdentityStoreImpl: IdentityStore {
+
+    private val identityLock = ReentrantReadWriteLock(true)
+    private var identity = Identity()
+
+    private val listenersLock = Any()
+    private val listeners: MutableSet<IdentityListener> = mutableSetOf()
+
+    override fun editIdentity(): IdentityStore.Editor {
+        val originalIdentity = getIdentity()
+        return object : IdentityStore.Editor {
+
+            private var userId: String? = originalIdentity.userId
+            private var deviceId: String? = originalIdentity.deviceId
+            private var userProperties: Map<String, Any?> = originalIdentity.userProperties
+
+            override fun setUserId(userId: String?): IdentityStore.Editor {
+                this.userId = userId
+                return this
+            }
+
+            override fun setDeviceId(deviceId: String?): IdentityStore.Editor {
+                this.deviceId = deviceId
+                return this
+            }
+
+            override fun setUserProperties(userProperties: Map<String, Any?>): IdentityStore.Editor {
+                this.userProperties = userProperties
+                return this
+            }
+
+            override fun updateUserProperties(actions: Map<String, Map<String, Any?>>): IdentityStore.Editor {
```

Comment:
We could make this more strict if `ID_OP_X` was an Enum and `actions: EnumMap<ID_OP, Map<String, Any?>>`. This could help with avoiding unhandled actions below.

Reply:
My feeling is that solution is too verbose for the added benefit. I prefer to keep things simple, and use strings.